sql: code generation for missing pg_catalog columns

mnovelodou opened the pull request:
Previously, keeping up with postgres by adding the same columns in pg_catalog
was a manual process
This was inadequate because from version to version the list of missing
columns might increase a lot
To address this, this patch closes the gap between postgres and cockroachdb
by adding the missing columns automatically with nil values

Release note: None

Fixes: #58001

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59302)
<!-- Reviewable:end -->

knz reviewed (commented):
What I am not seeing here is the distinction between empty strings and NULLs. Postgres uses empty strings for certain columns that have NOT NULL constraints. Arguably, we can't reasonably populate a NULL value inside a column where clients expect there will be no NULL value -- that would cause client errors.

Unfortunately, we did not build machinery inside cockroachdb to assert that virtual table generators don't populate NULL in NOT NULL columns. So if we do this by mistake, this will not be caught automatically.

What did you plan to do about this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-:-MRivIIq9S_ohdHspS1e:bh2novh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

Commit:
sql: code generation for missing pg_catalog columns

Previously, keeping up with postgres by adding the same columns in pg_catalog
was a manual process
This was inadequate because from version to version the list of missing
columns might increase a lot
To address this, this patch closes the gap between postgres and cockroachdb
by adding the missing columns automatically with nil values

Release note: None

Fixes: #58001

RichardJCai commented on pkg/sql/pg_metadata_test.go:
```diff
@@ -611,6 +746,239 @@ func getSortedDefKeys(tableDefs map[string]string) []string {
+
+// Visit implements ast.Visitor and sets the rules for detecting schema, matching schema with addRow calls and finding
+// which schemas have "return nil" at populate function
+func (v *pgCatalogCodeVisitor) Visit(node ast.Node) ast.Visitor {
```
Where does this get used?

mnovelodou commented on pkg/sql/pg_metadata_diff.go:
```diff
@@ -214,6 +215,53 @@ func (p PGMetadataTables) getUnimplementedTables(source PGMetadataTables) PGMeta
+	unimplementedColumns := make(PGMetadataTables)
+	for tableName, columns := range p {
+		if len(columns) == 0 {
+			//not implemented table
```
Done!
Nice catch!

mnovelodou commented on pkg/sql/pg_metadata_diff.go:
```diff
@@ -214,6 +215,53 @@ func (p PGMetadataTables) getUnimplementedTables(source PGMetadataTables) PGMeta
+	return unimplementedColumns
+}
+
+func (p PGMetadataTables) removeImplementedColumns(source PGMetadataTables) {
```
Done.

mnovelodou commented on pkg/sql/pg_metadata_diff.go:
```diff
@@ -226,3 +274,22 @@ func (c PGMetadataColumns) getUnimplementedTypes() map[oid.Oid]string {
+
+// TypeIsUnimplemented determine whether the type is implemented or not in
+// cockroachdb.
+func (t *PGMetadataColumnType) TypeIsUnimplemented() bool {
```
Done.

mnovelodou commented on pkg/sql/pg_metadata_test.go:
```diff
@@ -252,26 +258,56 @@ func fixConstants(t *testing.T, notImplemented PGMetadataTables) {
 
-// fixVtable adds missing table's create table constants.
-func fixVtable(t *testing.T, notImplemented PGMetadataTables) {
+// fixVtable adds missing table's create table constants
```
Sorry xD, Problems came by merge, but I tested the code, functionality is fine!

mnovelodou replied:
Done.

mnovelodou commented on pkg/sql/pg_metadata_test.go:
```diff
@@ -315,6 +397,53 @@ func fixPgCatalogGo(notImplemented PGMetadataTables) {
+
+				for _, columnName := range positions[currentPosition].missingColumns {
+					output.appendString(indentation)
+					output.appendString("tree.DNull,")
```
We already discussed this with @rafiss, we are doing as much as we can to improve compatibility, but addressing columns as not null will implicate to define a Zero value for that column that is interpreted as a zero value and not the actual value. That is the reason we are not even adding NOT NULL columns to the create tables.

I think the only way to add a NOT NULL column is to do it manually by implementing the column.

mnovelodou commented on pkg/sql/pg_metadata_test.go:
```diff
@@ -611,6 +746,239 @@ func getSortedDefKeys(tableDefs map[string]string) []string {
 	return keys
 }
 
+// goParsePgCatalogGo analyzes whether a virtualSchemaTable have a return nil populate function or maps addRow calls
```
re-worded

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_test.go, line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdtF-X2I1Vs6lO3ki-:-MXdtF-X2I1Vs6lO3ki0:bez4j5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L524)):*
> ```Go
> }
> 
> func (o outputFile) seekRelative(offset int64) {
> ```

Comment what this function does as well
___
*[pkg/sql/pg_metadata_test.go, line 828 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdssj0984Ro3z_OkHI:-MXdssj0984Ro3z_OkHJ:b-fgew6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L828)):*
> ```Go
> // Visit implements ast.Visitor and sets the rules for detecting schema, matching schema with addRow calls and finding
> // which schemas have "return nil" at populate function
> func (v *pgCatalogCodeVisitor) Visit(node ast.Node) ast.Visitor {
> ```

Appreciate the comments here since it really helps following the code but nit that they're missing periods 
___
*[pkg/sql/pg_metadata_test.go, line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdsR2c4ABiwjqVccXv:-MXdsR2c4ABiwjqVccXw:b-kke7k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L880)):*
> ```Go
> 			}
> 			if _, ok = v.pgCode.addRowPositions[v.schema]; !ok {
> 				v.pgCode.addRowPositions[v.schema] = make([]*addRowPosition, 0, 3)
> ```

This function is definitely going to be hard to understand in case someone new has to maintain it. I would try to avoid magic numbers, ie 3 here and perhaps use a constant.

Also in general I would emphasize that we over comment anything that might be slightly confusing here to make sure it's easy to follow.
___
*[pkg/sql/pg_metadata_test.go, line 891 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdrsMWAZSbUpXzTQzQ:-MXdrsMWAZSbUpXzTQzR:b-nxla2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L891)):*
> ```Go
> 			addRowList := v.pgCode.addRowPositions[v.schema]
> 			v.pgCode.addRowPositions[v.schema] = append(addRowList, addRow)
> 		case "makeAllRelationsVirtualTableWithDescriptorIDIndex":
> ```

I left a comment on your other PR about this but just to clarify, what would happen if this function for whatever reason was changed.
I know that this works as of now but I'm concerned about the robustness / maintainability here if any of this were to chase.


<!-- Sent from Reviewable.io -->

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_test.go, line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdtF-X2I1Vs6lO3ki-:-MYBUDDV1HRIa1GSBSf8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L524)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Comment what this function does as well
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 828 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdssj0984Ro3z_OkHI:-MYBUNFW2RD8c_cawlGr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L828)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Appreciate the comments here since it really helps following the code but nit that they're missing periods 
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdsR2c4ABiwjqVccXv:-MYCTvBx1uv06vX6U3yi:b-5vvz4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L880)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This function is definitely going to be hard to understand in case someone new has to maintain it. I would try to avoid magic numbers, ie 3 here and perhaps use a constant.

Also in general I would emphasize that we over comment anything that might be slightly confusing here to make sure it's easy to follow.
</blockquote></details>

I think for this particular code I noticed that most of the times, you call "addRow" 1 or 2 times, so I gave capacity of 3 to cover most of the cases when I find addRow and avoid growing the array inside the slice.

Below on missingColumns I am adding capacity of 5 thinking to cover most of the cases without growing the array of the slice (missing columns are the new amount of parameters you need in addRow calls).

I can add comments or constants for this particular capacity numbers or constants but I found kind of difficult to understand what other kind of comments you might need here, this is a code that is traversing a tree structure which nodes are tokens of the go code, every kind of token have it's own type (like variable or functions declarations, function calls, arguments, types, etc)

This code is traversing that tree structure to find specific portions of code that defines the populate on virtualSchemaTable objects to match the positions where the addRows to a tableName, so later in the code we can see if that table have new columns and edit the addRows calls.

I can add the explanation above but if you still find some difficulties to understand or find something confusing here, You need to point out what is the confusing part.
___
*[pkg/sql/pg_metadata_test.go, line 891 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdrsMWAZSbUpXzTQzQ:-MYCWxpSBBeLSc7qbcUP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L891)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I left a comment on your other PR about this but just to clarify, what would happen if this function for whatever reason was changed.
I know that this works as of now but I'm concerned about the robustness / maintainability here if any of this were to chase.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
Thanks I think this looks a lot better and is much easier to understand now.

Only comment nits to fix up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-:-MYlGjRT4NexdVFMMk6t:b-rmhcuf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_diff.go, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDQ4d4lkUAM3C4BSg:-MYlDQ4d4lkUAM3C4BSh:bz25web) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L226)):*
> ```Go
> 		for columnName, columnType := range columns {
> 			if columnType != nil {
> 				//dataType mismatch (Not a new column)
> ```

nit: space before starting sentence and period.

Should be // dataType mismatch (Not a new column).
___
*[pkg/sql/pg_metadata_diff.go, line 232 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDQqoChJgNHShJ9tY:-MYlDQqoChJgNHShJ9tZ:b8gem34) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L232)):*
> ```Go
> 			typeOid := oid.Oid(sourceType.Oid)
> 			if _, ok := types.OidToType[typeOid]; !ok || typeOid == oid.T_anyarray {
> 				//can't implement this column due to missing type
> ```

nit: space and period.
___
*[pkg/sql/pg_metadata_diff.go, line 242 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDXvQDhHKnMf2zjvL:-MYlDXvQDhHKnMf2zjvM:bldqbl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L242)):*
> ```Go
> 
> // removeImplementedColumns removes diff columns that are marked as expected
> // diff (or unimplmented column) but is already implemented in CRDB
> ```

nit: typo in uimplmented and period to finish comment.
___
*[pkg/sql/pg_metadata_diff.go, line 255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDdLd2Jux44E3jaGh:-MYlDdLd2Jux44E3jaGi:bnq9baj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L255)):*
> ```Go
> 			}
> 			if columnType != nil {
> 				// type diff
> ```

We can just remove this.
___
*[pkg/sql/pg_metadata_diff.go, line 287 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDmxd8wmSYP3r5cl0:-MYlDmxd8wmSYP3r5cl1:b-lfazfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L287)):*
> ```Go
> }
> 
> // IsImplemented determine whether the type is implemented or not in
> ```

nit: determines
___
*[pkg/sql/pg_metadata_test.go, line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdsR2c4ABiwjqVccXv:-MYlGWf46tgu4ydMX1m5:b-kufkx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L880)):*
<details><summary><i>Previously, mnovelodou (Miguel Novelo) wrote…</i></summary><blockquote>

I think for this particular code I noticed that most of the times, you call "addRow" 1 or 2 times, so I gave capacity of 3 to cover most of the cases when I find addRow and avoid growing the array inside the slice.

Below on missingColumns I am adding capacity of 5 thinking to cover most of the cases without growing the array of the slice (missing columns are the new amount of parameters you need in addRow calls).

I can add comments or constants for this particular capacity numbers or constants but I found kind of difficult to understand what other kind of comments you might need here, this is a code that is traversing a tree structure which nodes are tokens of the go code, every kind of token have it's own type (like variable or functions declarations, function calls, arguments, types, etc)

This code is traversing that tree structure to find specific portions of code that defines the populate on virtualSchemaTable objects to match the positions where the addRows to a tableName, so later in the code we can see if that table have new columns and edit the addRows calls.

I can add the explanation above but if you still find some difficulties to understand or find something confusing here, You need to point out what is the confusing part.
</blockquote></details>

I see, thanks.
___
*[pkg/sql/pg_metadata_test.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlE1LsDZUV2YLkZPhl:-MYlE1LsDZUV2YLkZPhm:b-7u3vpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L104)):*
> ```Go
> 	unimplemented: true,
> }
> ```

Remove blank line here.
___
*[pkg/sql/pg_metadata_test.go, line 371 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlErPI5_9D1FnlY244:-MYlErPI5_9D1FnlY245:b-1v3kle) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L371)):*
> ```Go
> 
> // getMissingColumnsText creates the text used by a vTable constant to add the
> // newer (or missing) columns.
> ```

nit: newer -> new
___
*[pkg/sql/pg_metadata_test.go, line 440 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlFAZd4EuGkHIPr82X:-MYlFAZd4EuGkHIPr82Y:b-lewylb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L440)):*
> ```Go
> 	rewriteFile(pgCatalogGo, func(input *os.File, output outputFile) {
> 		reader := bufio.NewScanner(input)
> 		soFar := 0
> ```

I think soFar should be renamed to scannedUntil.

Thanks for the comments here though definitely easier to understand.
___
*[pkg/sql/pg_metadata_test.go, line 874 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlFcbWBNiNKyf203KF:-MYlFcbWBNiNKyf203KG:b-8so11) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L874)):*
> ```Go
> 	fixableTables   map[string]struct{}
> 	addRowPositions mappedRowPositions
> 	schemaParam     int // Index where we expect to find schema at makeAllRelationsVirtualTableWithDescriptorIDIndex
> ```

nit: period


<!-- Sent from Reviewable.io -->

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_diff.go, line 244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MWVdC1mPKbfH0m5OvVI-r1-244:-MYpPGbi3uh8M7cRxd0c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdab5082a76859f5c706ef44116ccd86b23f2c30/pkg/sql/pg_metadata_diff.go#L244)):*
<details><summary><i>Previously, mnovelodou (Miguel Novelo) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_diff.go, line 255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlDdLd2Jux44E3jaGh:-MYpPRJf1dQ8RPm2jiSn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_diff.go#L255)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We can just remove this.
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MXdsR2c4ABiwjqVccXv:-MYpP_5wB44bFQctmqnk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/acdf46da9546b63c5815fe11d2dac15abb095f83/pkg/sql/pg_metadata_test.go#L880)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I see, thanks.
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlE1LsDZUV2YLkZPhl:-MYpQIrW-LRU5NjHRaZE:brrqmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L104)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Remove blank line here.
</blockquote></details>

The blank line is on purpose to separate this objects and have the linters pass
![Screen Shot 2021-04-21 at 12.16.47.png](https://files.reviewableusercontent.io/images/reviewable/74206951/PjQPAAtJelGa19YQNKZndHWa/Screen%20Shot%202021-04-21%20at%2012.16.47.png)
___
*[pkg/sql/pg_metadata_test.go, line 440 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlFAZd4EuGkHIPr82X:-MYpQ5tM2TiXrfsh2xCY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L440)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think soFar should be renamed to scannedUntil.

Thanks for the comments here though definitely easier to understand.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_test.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlE1LsDZUV2YLkZPhl:-MYpsaWdA67Z49_AHhD_:b-hsemka) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L104)):*
<details><summary><i>Previously, mnovelodou (Miguel Novelo) wrote…</i></summary><blockquote>

The blank line is on purpose to separate this objects and have the linters pass
![Screen Shot 2021-04-21 at 12.16.47.png](https://files.reviewableusercontent.io/images/reviewable/74206951/PjQPAAtJelGa19YQNKZndHWa/Screen%20Shot%202021-04-21%20at%2012.16.47.png)

</blockquote></details>

Ah I see.
___
*[pkg/sql/pg_metadata_test.go, line 1144 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYptMuVEH3i_z_mONnH:-MYptMuW-1HQtcYjjbSw:b-cuz5d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b144de3364401c9a9d034c69dff5d21eda6e648b/pkg/sql/pg_metadata_test.go#L1144)):*
> ```Go
> 	read, err := srcFile.Read(bytes)
> 	if err != nil {
> 		return "", nil
> ```

I think the linter is failing because we're not returning an error here.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
LGTM after last fix.

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @solongordon)

___
*[pkg/sql/pg_metadata_test.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYlE1LsDZUV2YLkZPhl:-MYq4DfKA8nhGT8e3sYG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b38ee1dcb27ee25682a8fb4a5808cda1844b7f0e/pkg/sql/pg_metadata_test.go#L104)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Ah I see.
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 1144 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59302#-MYptMuVEH3i_z_mONnH:-MYq5awu-aQE4Ylsvyfw:b7ajauf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b144de3364401c9a9d034c69dff5d21eda6e648b/pkg/sql/pg_metadata_test.go#L1144)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think the linter is failing because we're not returning an error here.
</blockquote></details>

thanks, nice catch!


<!-- Sent from Reviewable.io -->

RichardJCai commented:
bors r+
